### PR TITLE
[Don't merge] add: poc for support freeU.

### PR DESCRIPTION
**Pull request by sayakpaul**
To continue jamming on https://github.com/huggingface/diffusers/pull/5164.

### CODE REVIEW

File: src/diffusers/models/unet_2d_blocks.py [hunk 1]
```diff
@@ -2194,11 +2194,24 @@ def forward(
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
     ):
         lora_scale = cross_attention_kwargs.get("scale", 1.0) if cross_attention_kwargs is not None else 1.0
+        is_freeu_enabled = getattr(self.config, "s1", None) and getattr(self.config, "s2", None) and getattr(self.config, "b1", None), getattr(self.config, "b2", None)
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2197]:
```suggestion
        is_freeu_enabled = getattr(self.config, "s1", None) and getattr(self.config, "s2", None) and getattr(self.config, "b1", None) and getattr(self.config, "b2", None)  
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 2]
```diff
@@ -2293,11 +2306,32 @@ def __init__(
...
-    def forward(self, hidden_states, res_hidden_states_tuple, temb=None, upsample_size=None, scale: float = 1.0):
+    def forward(
+        self,
+        hidden_states,
+        res_hidden_states_tuple,
+        temb=None,
+        upsample_size=None,
+        scale: float = 1.0,
+    ):
+        is_freeu_enabled = getattr(self.config, "s1", None) and getattr(self.config, "s2", None) and getattr(self.config, "b1", None), getattr(self.config, "b2", None)
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2317]:
```suggestion
        is_freeu_enabled = getattr(self.config, "s1", None) and getattr(self.config, "s2", None) and getattr(self.config, "b1", None) and getattr(self.config, "b2", None)  
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -533,6 +533,24 @@ def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype
         # scale the initial noise by the standard deviation required by the scheduler
         latents = latents * self.scheduler.init_noise_sigma
         return latents
+    
+    def enable_freeu(self, **freeu_kwargs):
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:537]:
```suggestion
    def enable_freeu(self, s1=0.9, s2=0.2, b1=1.2, b2=1.4):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 4]
```diff
@@ -2293,11 +2306,32 @@ def __init__(
...
+            if is_freeu_enabled:
+                # --------------- FreeU code -----------------------
+                # Only operate on the first two stages
+                # `fourier_filter` needs to be implemented.
+                if hidden_states.shape[1] == 1280:
+                    hidden_states[:, :640] = hidden_states[:, :640] * self.config["b1"]
+                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.config["s1"])
+                if hidden_states.shape[1] == 640:
+                    hidden_states[:, :320] = hidden_states[:, :320] * self.config["b2"]
+                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.config["s2"])
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2333]:
```suggestion
                if self.resolution_idx == 0:  # add this to init
                    hidden_states[:, :hidden_states.shape[-1] // 2] = hidden_states[:, :hidden_states.shape[-1] // 2] * self.config["b1"]
                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.config["s1"])
                elif self.resolution_idx == 1:  # add this to init
                    hidden_states[:, :hidden_states.shape[-1] // 2] = hidden_states[:, :hidden_states.shape[-1] // 2] * self.config["b2"]
                    res_hidden_states = fourier_filter(res_hidden_states, threshold=1, scale=self.config["s2"])
```

Not a huge fan of hard-coding values in here - can we instead pass the resolution index to the block to make it more explicit?

Comment (sayakpaul) [reply src/diffusers/models/unet_2d_blocks.py:2333]: `resolution_idx` does not seem naturally connecting to me. What does it mean?

Comment (patrickvonplaten) [reply src/diffusers/models/unet_2d_blocks.py:2333]: Yeah maybe `block_idx` is better

**Comment by sayakpaul** [comment]
Closing in favor of the original PR.

**Pull request closed without merging**